src/Language/Haskell/Liquid/GHC/Interface.hs

```diff
@@ -72,284 +79,339 @@ import Language.Fixpoint.Utils.Files
+type SpecEnv = ModuleEnv (ModName, Ms.BareSpec)
 
-derivedVs :: CoreProgram -> DFunId -> [Id]
-derivedVs cbs fd = concatMap bindersOf cbs' ++ deps
+processModules :: Config -> Either Error LogicMap -> [FilePath] -> DepGraph
+               -> ModuleGraph
+               -> Ghc [GhcInfo]
+processModules cfg logicMap tgtFiles depGraph homeModules =
+  catMaybes . snd <$> mapAccumM go emptyModuleEnv homeModules
   where
-    cbs'           = filter f cbs
-    f (NonRec x _) = eqFd x
-    f (Rec xes)    = any eqFd (fst <$> xes)
-    eqFd x         = varName x == varName fd
-    deps           = concatMap unfoldDep unfolds
-    unfolds        = unfoldingInfo . idInfo <$> concatMap bindersOf cbs'
-
-unfoldDep :: Unfolding -> [Id]
-unfoldDep (DFunUnfolding _ _ e)         = concatMap exprDep e
-unfoldDep (CoreUnfolding {uf_tmpl = e}) = exprDep e
-unfoldDep _                             = []
+    go = processModule cfg logicMap (S.fromList tgtFiles) depGraph
+  
+processModule :: Config -> Either Error LogicMap -> S.HashSet FilePath -> DepGraph
+              -> SpecEnv -> ModSummary
+              -> Ghc (SpecEnv, Maybe GhcInfo)
+processModule cfg logicMap tgtFiles depGraph specEnv modSummary = do
+  let mod              = ms_mod modSummary
+  _                   <- liftIO $ whenLoud $ putStrLn $ "Module: " ++ showPpr (moduleName mod)
+  file                <- liftIO $ canonicalizePath $ modSummaryHsFile modSummary
+  _                   <- loadDependenciesOf $ moduleName mod
+  parsed              <- parseModule $ keepRawTokenStream modSummary
+  typechecked         <- typecheckModule $ ignoreInline parsed
+  _                   <- loadModule typechecked
+  let specComments     = getSpecComments parsed
+  (modName, bareSpec) <- either throw return $ hsSpecificationP (moduleName mod) specComments
+  let specEnv'         = extendModuleEnv specEnv mod (modName, noTerm bareSpec)
+  (specEnv', ) <$> if not (file `S.member` tgtFiles)
+    then return Nothing
+    else Just <$> processTargetModule cfg logicMap depGraph specEnv file typechecked bareSpec
+
+keepRawTokenStream :: ModSummary -> ModSummary
+keepRawTokenStream modSummary = modSummary
+  { ms_hspp_opts = ms_hspp_opts modSummary `gopt_set` Opt_KeepRawTokenStream }
+
+loadDependenciesOf :: ModuleName -> Ghc ()
+loadDependenciesOf modName = do
+  loadResult <- load $ LoadDependenciesOf modName
+  when (failed loadResult) $ liftIO $ throwGhcExceptionIO $ ProgramError $
+   "Failed to load dependencies of module " ++ showPpr modName
+
+getSpecComments :: ParsedModule -> [(SourcePos, String)]
+getSpecComments parsed = mapMaybe getSpecComment comments
+  where
+    comments = concat $ M.elems $ snd $ pm_annotations parsed
 
-exprDep :: CoreExpr -> [Id]
-exprDep = freeVars S.empty
+getSpecComment :: GHC.Located AnnotationComment -> Maybe (SourcePos, String)
+getSpecComment (GHC.L span (AnnBlockComment text))
+  | length text > 2 && isPrefixOf "{-@" text && isSuffixOf "@-}" text =
```

Comment:
Seems redundant given the other checks ...

On Friday, April 29, 2016, Eric Seidel <notifications@github.com> wrote:

> In src/Language/Haskell/Liquid/GHC/Interface.hs
> <https://github.com/ucsd-progsys/liquidhaskell/pull/698#discussion_r61606539>
> :
>
> >
> > -exprDep :: CoreExpr -> [Id]
> > -exprDep = freeVars S.empty
> > +getSpecComment :: GHC.Located AnnotationComment -> Maybe (SourcePos, String)
> > +getSpecComment (GHC.L span (AnnBlockComment text))
> > +  | length text > 2 && isPrefixOf "{-@" text && isSuffixOf "@-}" text =
>
> Do we actually need the length > 2 test?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/ucsd-progsys/liquidhaskell/pull/698/files/891243874be9710f18272a399aedadda52c889be#r61606539>
>